### PR TITLE
PMM-10948 add "source" label gRPC prometheus metrics

### PR DESCRIPTION
**Pull request by YaroslavPodorvanov**
[PMM-10948](https://jira.percona.com/browse/PMM-10948) ([PMM-10541](https://jira.percona.com/browse/PMM-10541))

https://github.com/percona/pmm/pull/1370

https://github.com/Percona-Lab/go-grpc-prometheus/pull/3 better

### CODE REVIEW

File: source_type.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	"context"
+	"strconv"
+)
+
+type sourceContextKeyType struct{}
+
+// sourceContextKey unique key for all packages
+var sourceContextKey sourceContextKeyType
+
+type Source int
```
Comment (ritbl) [inline source_type.go:13]: source sounds to me quite generic, how about something like `caller_origin`?

File: server_metrics.go [hunk 2]
```diff
@@ -121,7 +121,7 @@ func (m *ServerMetrics) StreamServerInterceptor() func(srv interface{}, ss grpc.
 		monitor := newServerReporter(m, streamRPCType(info), info.FullMethod)
 		err := handler(srv, &monitoredServerStream{ss, monitor})
 		st, _ := grpcstatus.FromError(err)
-		monitor.Handled(st.Code())
+		monitor.Handled(st.Code(), Internal)
```
Comment (ritbl) [inline server_metrics.go:124]: why it is internal?

File: source_type.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+// SetSourceToCtx returns derived context with source
+func SetSourceToCtx(ctx context.Context, source Source) context.Context {
+	return context.WithValue(ctx, sourceContextKey, source)
+}
+
+// GetSourceFromCtx returns source for given context
+func GetSourceFromCtx(ctx context.Context) Source {
+	v := ctx.Value(sourceContextKey)
+	if v == nil {
+		return Internal
```
Comment (ritbl) [inline source_type.go:43]: Why it is internal if key is not in cotext?

File: source_type.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	}
+}
+
+// SetSourceToCtx returns derived context with source
+func SetSourceToCtx(ctx context.Context, source Source) context.Context {
+	return context.WithValue(ctx, sourceContextKey, source)
+}
+
+// GetSourceFromCtx returns source for given context
+func GetSourceFromCtx(ctx context.Context) Source {
```
Comment (ritbl) [inline source_type.go:40]: I think it should return ref type, and `nil` in case value is not in context.

File: server_test.go [hunk 5]
```diff
@@ -120,10 +120,10 @@ func (s *ServerInterceptorTestSuite) TestRegisterPresetsStuff() {
 		{"grpc_server_msg_sent_total", []string{"mwitkow.testproto.TestService", "PingEmpty", "unary"}},
 		{"grpc_server_handling_seconds_sum", []string{"mwitkow.testproto.TestService", "PingEmpty", "unary"}},
 		{"grpc_server_handling_seconds_count", []string{"mwitkow.testproto.TestService", "PingList", "server_stream"}},
-		{"grpc_server_handled_total", []string{"mwitkow.testproto.TestService", "PingList", "server_stream", "OutOfRange"}},
-		{"grpc_server_handled_total", []string{"mwitkow.testproto.TestService", "PingList", "server_stream", "Aborted"}},
-		{"grpc_server_handled_total", []string{"mwitkow.testproto.TestService", "PingEmpty", "unary", "FailedPrecondition"}},
-		{"grpc_server_handled_total", []string{"mwitkow.testproto.TestService", "PingEmpty", "unary", "ResourceExhausted"}},
+		{"grpc_server_handled_total", []string{"mwitkow.testproto.TestService", "PingList", "server_stream", "OutOfRange", "internal"}},
```
Comment (ritbl) [inline server_test.go:123]: by default, I think, we should not have this label

**Comment by YaroslavPodorvanov** [comment]
Draft after discussions with @ritbl and @artemgavrilov to try use https://github.com/grpc-ecosystem/go-grpc-middleware/tree/v2

**Comment by YaroslavPodorvanov** [comment]
https://github.com/Percona-Lab/go-grpc-prometheus/pull/3 better

**Pull request closed without merging**